### DISCUSSION TITLE
Why can model_pred simply subtract noise to fit targets=x0?

### DISCUSSION THREAD

**Discussion by zhangqizky** [huggingface/diffusers#6622] [Q&A]
https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/train_text_to_image_sdxl.py#L1079
here, when prediction_type=="sample", means UNet predict x0, and the targets also set to x0, but why the model_pred need to subtract noise? I don't much undstand here... Could anybody help to explain?

**Comment by sayakpaul** [comment]
I don't fully understand the question. How would you get the target sample otherwise?

**Reply by zhangqizky** [reply]
I'm not sure otherwise it should be. I just don't understand, what "model_pred = model_pred - noise " phsically mean? It should mean "predicted-x0", but the training target is "GT-x0", after training, the model ouput should be "predicted-x0", I don't understand why need to subtract noise. I think should be sth like this , not simply subtract? I didn't find any explain in ddpm-paper aboud "model_pred - noise", so I got confused....
<img width="401" alt="image" src="https://github.com/huggingface/diffusers/assets/23207132/1738e577-c282-47c8-893b-f678cee3e86b">

**Comment by christopher-beckham** [comment]
@zhangqizky @sayakpaul I don't really understand it either. Based on these two lines:

- https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/train_text_to_image_sdxl.py#L1126
- https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/train_text_to_image_sdxl.py#L1064

we have `model_pred = model_pred - epsilon`, which I will just re-write to `model_pred = unet(x_noised) - epsilon`, where `x_noised = add_noise(x, sigma, epsilon)` (which would be e.g. `x + sigma*epsilon` for Euler schedulers) and `epsilon` is the original _prescaled_ noise, i.e. `epsilon ~ N(0,1)`. Since we're using a mean squared error loss,  under a prediction type of "sample" it would obviously be:

```
loss = || model_pred - x ||^2
     = || (unet(x_noised) - epsilon) - x||^{2}
```

which means that `unet(x_noised) = x + epsilon` at optimality (to make the loss zero). This is an odd and I've never seen this formulation before. Either:
- (1) The `unet` should directly predict the original sample, i.e. `unet(x_noised) = x`, then the loss is `||unet(x_noised) - x||^2`. This would truly correspond to the prediction type being "sample".
- (2) or you define the denoised sample in terms of unet being an epsilon predictor, i.e. `x0_predicted = x_noised - sigma*unet(x_noised)`, and then do the minimise loss `||x0_predicted - x||^2`. However this is sort of a moot point since it implies the prediction type should be "epsilon" and we're not doing that anyway.

It looks like most schedulers don't support `prediction_type == 'sample'`, but for the ones that do it is assumed that (1) is the case, e.g. see [here](https://github.com/huggingface/diffusers/blob/fe5f035f797a5fa663a98030c9d0ec2f982cd09d/src/diffusers/schedulers/scheduling_ddpm_flax.py#L247) and [here](https://github.com/huggingface/diffusers/blob/fe5f035f797a5fa663a98030c9d0ec2f982cd09d/src/diffusers/schedulers/scheduling_ddim_flax.py#L262).

**Comment by zhangqizky** [comment]
您的邮件已收到，我会尽快回复，谢谢。张琦

**Comment by sasasqt** [comment]
any updates on this? @bghira

**Comment by zhangqizky** [comment]
您的邮件已收到，我会尽快回复，谢谢。张琦
